accepted/2020/better-obsoletion/securestring-obsoletion.md

```diff
@@ -0,0 +1,237 @@
+
+The type also intentionally does not give callers access to the raw backing span. This is because how the backing data is stored is strictly an implementation detail (it might not even be in-proc!), and exposing the span would leak that detail into the public API surface. The `CopyTo` method is the only way to get the raw contents. This also means that such buffers are immutable once created.
+
+__We should take care not to position this "sensitive buffer" type as a security feature for the purpose of achieving compliance.__ It's intended solely for code hygiene purposes and to avoid accidental disclosure. Scenarios like achieving PCI compliance require more fundamental runtime work and security audits, such as those mentioned at https://github.com/dotnet/runtime/issues/10480.
+
+This proposal intentionally excludes `ToString` and other methods which might unshroud the buffer. `ToString` is called frequently by library code as part of logging, and we always want unshrouding to be a _deliberate_ action. We want the caller to maintain full control over the unshrouded buffer where the contents are written.
+
+> This does not exactly preclude the introduction of an `ToArrayIMeanItReally` API. Such an API call would be clearly deliberate. The real difficulity is that the caller may be opinionated as to how the array is created (POH? normal heap but pinned? pooled?), and our opinionated `ToArray` behavior may not match the caller's opinionated behavior. Having a single `CopyTo` method allows the caller to maintain full control over how the buffer is allocated, and we simply populate the buffer. But if this ends up being a usage hurdle, we can always reconsider this decision.
+
+If auditors want to monitor "unshroud" operations, they can look for calls to `CopyTo`.
+
+### Best practices for implementation
+
+If the backing buffer for the sensitive data is stored in-proc, we should attempt to use a separate unmanaged heap for storing these values, and we should not hand out direct references to this heap. This helps mitigate the possibility of sensitive data being leaked due to corruption of the normal heap. This concept is [already in use by OpenSSL](https://www.openssl.org/docs/man1.1.0/man3/CRYPTO_secure_malloc_initialized.html) and a few high-sensitivity Windows services. If using a separate heap is not feasible, we should attempt to use .NET's pinned object heap.
+
+The backing buffer may be (but is not required to be) enciphered or marked do-not-dump. In practice, we will probably encipher data when we can to avoid accidental leakage of the sensitive data. We will probably not mark the pages as do-not-dump due to the difficulty of bookkeeping and the fact that both Linux and Windows primarily intend do-not-dump regions to be for large data sections (such as texture data for games), and they institute an upper bound on the number of pages which may be annotated as do-not-dump.
+
+## New overloads on existing APIs
```

Comment:
what about the IDataProtector api's?
maybe I'm storing some "sensitive" data like
- passwords for unlocking PFX cert containers
- api token's that are transported encrypted from a (more) secured backend

having the 1 byte at a time would also be very nice to limit what can be leaked from memory dumping attempts

Reply:
i always try to defense in depth whenever possible
we never know when we end up with the next version of Heartbleed
so the least amount of time data is spent (partially) unencrypted the better